### PR TITLE
contracts-bedrock: update weth98

### PR DESCRIPTION
**Pull request by tynes**
**Description**

The custom gas token project is moving away from using
the original weth9 code as a predeploy to using the
weth98 code as a predeploy. As part of this, the ability
to override the name and symbol are required for chains
that use a custom gas token. There is no way to
mark an inline declaration as `virtual`, it requires
moving to a `function` syntax. Definitely less clean
looking but necessary to support the override ability.

The implementation of the new WETH predeploy is as follows:

```solidity
contract WETH is WETH98 {
    /// @notice Returns the name of the token from the L1Block contract
    function name() external view override returns (string memory) {
        return string.concat("Wrapped ", L1Block(Predeploys.L1_BLOCK_ATTRIBUTES).gasPayingTokenName());
    }

    /// @notice Returns the symbol of the token from the L1Block contract
    function symbol() external view override returns (string memory) {
        return string.concat("W", L1Block(Predeploys.L1_BLOCK_ATTRIBUTES).gasPayingTokenSymbol());
    }
}
```

Where when not in custom gas token mode, it with return `Wrapped Ether`
and `WETH`, so it is backwards compatible.

This commit also adds a fallback function since the original weth9 has
a fallback function but this one does not. While the receive function
can work, adding the fallback function ensures that the same behavior
as weth9 is achieved when making a call that includes calldata + value.

## Summary by CodeRabbit

- **Refactor**
    - Updated the `name` and `symbol` properties in WETH contracts to be accessible via functions, enhancing contract interaction.
    - Added a new `fallback` function to automatically handle deposits, improving usability.
    - Updated the visibility specifier of functions in the WETH contract to `view`, indicating read-only operations.
    - Updated the functionality in op-bindings related to the `DelayedWETHDeployedBin` variable value.
    - Added a payable fallback function in the DelayedWETH contract, altering behavior for incoming Ether transactions.
    - Added a new entry for a payable fallback function in the WETH98 contract JSON file.

### CODE REVIEW

**Review by refcell — APPROVED** [review]

**Comment by refcell** [comment]
Looks good, was just about to comment on the fallback rename

File: packages/contracts-bedrock/src/dispute/weth/WETH98.sol [hunk 1]
```diff
@@ -36,6 +40,11 @@ contract WETH98 is IWETH {
         deposit();
     }
 
+    /// @notice Pipes to deposit.
+    fallback() external payable {
+        deposit();
+    }
+
```
Comment (clabby) [inline packages/contracts-bedrock/src/dispute/weth/WETH98.sol:41]:
```suggestion
```

**Comment by tynes** [comment]
Will push fixes for bindings/snapshots

File: packages/contracts-bedrock/src/dispute/weth/WETH98.sol [hunk 2]
```diff
@@ -24,8 +24,12 @@ import { IWETH } from "src/dispute/interfaces/IWETH.sol";
 /// @title WETH98
 /// @notice WETH98 is a version of WETH9 upgraded for Solidity 0.8.x.
 contract WETH98 is IWETH {
-    string public constant name = "Wrapped Ether";
-    string public constant symbol = "WETH";
+    function name() external view virtual override returns (string memory) {
```
Comment (clabby) [inline packages/contracts-bedrock/src/dispute/weth/WETH98.sol]: See why you put these here given the old location of the constants, but can we place functions in their proper location in the file?

File: packages/contracts-bedrock/src/dispute/weth/WETH98.sol (hunk 1 above)
Comment (clabby) [reply packages/contracts-bedrock/src/dispute/weth/WETH98.sol:41]: We don't need a fallback here; `receive` will be triggered when positive `msg.value` is sent to the contract.

**Review by clabby — CHANGES_REQUESTED** [review]

File: packages/contracts-bedrock/src/dispute/weth/WETH98.sol (hunk 2 above)
Comment (tynes) [reply packages/contracts-bedrock/src/dispute/weth/WETH98.sol]: sure thing

File: packages/contracts-bedrock/src/dispute/weth/WETH98.sol (hunk 1 above)
Comment (tynes) [reply packages/contracts-bedrock/src/dispute/weth/WETH98.sol:41]: Without the fallback defined explicitly, it impacts the ABI in a non backwards compat way. Is it the fact that fallback accepts `calldata.length > 0` where receive accepts `calldata.length == 0`?

File: packages/contracts-bedrock/src/dispute/weth/WETH98.sol (hunk 2 above)
Comment (tynes) [reply packages/contracts-bedrock/src/dispute/weth/WETH98.sol]: i updated this

File: packages/contracts-bedrock/src/dispute/weth/WETH98.sol (hunk 1 above)
Comment (tynes) [reply packages/contracts-bedrock/src/dispute/weth/WETH98.sol:41]: You can see this change in the e2e tests to handle the non backwards compat: https://github.com/ethereum-optimism/optimism/pull/10090

Comment (clabby) [reply packages/contracts-bedrock/src/dispute/weth/WETH98.sol:41]: > Is it the fact that fallback accepts calldata.length > 0 where receive accepts calldata.length == 0?

Yep, `receive` gets hit if `eq(calldatasize(), 0)` + `gt(callvalue(), 0)` & `fallback` gets hit if `gt(calldatasize(), 0)` + no selectors match. But shouldn't this be the intended behavior? If I'm sending ETH to the `WETH98` contract, I send as a regular transfer (commonly `eq(calldatasize(), 0)`) or through a `payable` function. Is there a usecase for being able to send ETH through the fallback by providing useless calldata?

Comment (tynes) [reply packages/contracts-bedrock/src/dispute/weth/WETH98.sol:41]: The usecase is backwards compatibility with WETH9 as it supports this feature. Even though its seemingly useless, it preserves the ABI

Comment (clabby) [reply packages/contracts-bedrock/src/dispute/weth/WETH98.sol:41]: Okay, no worries. Let's keep it.

**Review by clabby — APPROVED** [review]

**Pull request merged**